### PR TITLE
First step to transition to qualified references

### PR DESCRIPTION
**Pull request by benharsh**
Co-authored with @mppf 
Reviewed by @vasslitvinov , @noakesmichael , @ronawho 

This is the first step towards a representation of references that does not
rely on embedding ref-ness into a symbol's Chapel type. Instead, symbols have
a QualifiedType that represents a varaible 'kind' like 'const', 'var', or
'ref'.

While there are a number of entries in the 'Qualifier' enum, only a subset are
currently relied upon:
- QUAL_UNKNOWN
- QUAL_REF
- QUAL_CONST_REF
- QUAL_WIDE_REF
- QUAL_VAL

Over time others may be introduced and utilized properly, but for this
initial step it was useful to keep the changes as simple as possible.
I have added a number of 'BHARSH TODO' comments to keep track
of future work required to convert the compiler entirely over to using
qualified references. I plan to continue working on these TODOs for
the upcoming release.

New in the compiler
-------------------

New utility functions:
- isMoveOrAssign
- isDerefMove (moving from a reference into a var)

Every BaseAST now supports the following queries that return a boolean:
- isRef()
- isWideRef()
- isRefOrWideRef()

These methods rely on the QualifiedType of symbol or expression they
represent. For now, they will also consider the Chapel type and look for
FLAG_REF and FLAG_WIDE_REF when appropriate. Eventually they will only
consider the Qualifier.

This change also introduces the primitive PRIM_SET_REFERENCE, which will
ultimately replace PRIM_ADDR_OF in the QualifiedType world.

The pass 'resolveIntents' is currently also responsible for setting the
correct Qualifier for variables.

Codegen
-------

While it would have been nice to use QualifiedTypes for the entirety of
codegen, in the interest of time it was most expedient to change symbols with
a "ref" Qualifier to have a reference type at the start of codegen. This made
it easy to get LLVM operational. That said, much of codegen has been changed
to use 'isRef', or something similar, instead of directly looking at flags on
the type's symbol.

There are many changes to codegen, most of them adding dereferences or
addrOfs when necessary. With QualifiedTypes, the compiler is less likely to
introduce reference or dereference temporaries, and codegen will pick up the
slack. For example, one can now dereference simply by PRIM_MOVE-ing into a
non-reference variable. A number of issues with codegen were exposed because
of arguments with a "ref" intent that did not have a reference type, and
therefore were not correctly identified as references by codegen.

Primitive Calls
---------------

All type functions for primitives now return QualifiedTypes. 'typeInfo'
itself still returns the raw Chapel type, but one can use "call->qualType()" to
retrieve the QualifiedType of a call.

There is more work to be done to utilize QualifiedTypes correctly for
primitive calls, but would require more work in earlier passes to be
completed first.

Verify
------

When the "--verify" flag is thrown, every pass after 'resolveIntents' will be
checked to prevent symbols with the QUAL_UNKNOWN intent. By this point, we
should know the kind of a variable.

Optimizations
-------------

Many of our optimization passes were touched as a result of this work. Some
changes are simply switching to 'isRef', others are more complicated. For
example:
- Needed to understand a dereferencing-move (isDerefMove)
- Needed to handle PRIM_ASSIGN *and* PRIM_MOVE
- Needed to handle PRIM_SET_REFERENCE like PRIM_ADDR_OF

While single-locale performance looks to be on-par with master prior to this
change, I suspect there are issues with copyPropagation and refPropagation. I
have also observed scalarReplace being less effective for some tests (e.g.,
promoted stream).

Regressions
-----------

The following tests demonstrated comm-count regressions and have had their
.good files changed:
  - multilocale/deitz/needMultiLocales/streamCommCheck
  - multilocale/deitz/needMultiLocales/diagnostics/test_remote_value_forwarding3
  - performance/hilde/userSync
  - studies/hpcc/CommDiags/stream_commDiags
  - studies/ssca2/performance/SSCA2_commDiags

I believe these regressions to be related to two changes:
- RVF correctly identifying references
- fewer unnecessary dereferences when passing a variable to a function

Going forward I will attempt to bring these comm counts back down.

Miscellaneous
-------------

While building promotion wrappers, we now make an effort to place the "p_i_"
indices within the body of the for loop we're creating. Instead of creating a
VarSymbol, we insert UnresolvedSymExprs into the relevant calls and rely on
'destructureIndices' in build.cpp to create the VarSymbol for us. Afterwards,
we walk the newly-created wrapper and convert any UnresolvedSymExprs we
inserted into SymExprs to the "p_i_" variable.

The dsiAccess for BlockArr previously had a dereference erroneously inserted
before some calls to the function, which resulted in a local copy of the
index the user wants to access. With this work the dereference goes away and
if a remote index is passed, the local block inside dsiAccess will be
violated. I attempted to use the 'in' and 'const in' intents, but both failed
to work. I intend to look into this further post-merge. For now, I create a
local variable and assign into it with the formal.

Most AST printing functions (e.g., print_view, list_view) have been updated
to display the Qualifier for a symbol. As a team we may want to change the
current format based on style preferences.

### CODE REVIEW

File: compiler/AST/astutil.cpp [hunk 1]
```diff
@@ -655,36 +679,52 @@ bool isTypeExpr(Expr* expr)
   return false;
 }
 
+static void pruneVisit(TypeSymbol* ts, Vec<FnSymbol*>& fns, Vec<TypeSymbol*>& types);
```
Comment (noakesmichael) [inline compiler/AST/astutil.cpp]: There seem to me to be 3 variations of pruneVisit().  Two are named pruneVisit and one is named printVisitFn().
Inconsistentcy or deliberate?

File: compiler/AST/expr.cpp [hunk 2]
```diff
@@ -141,6 +141,60 @@ const char* DefExpr::name() const {
...
+  if (CallExpr* call = toCallExpr(e))
+    if (call->isPrimitive(PRIM_ADDR_OF))
+      if (SymExpr* se = toSymExpr(call->get(1)))
+        if (se->var->qualType().isWideRef())
+          return true;
+
+  return false;
+}
+
+bool Expr::isRef() {
```
Comment (noakesmichael) [inline compiler/AST/expr.cpp]: Inconsistent whitespace after "if".   FWIW I prefer 1 space.

**Comment by noakesmichael** [comment]
But not catching TypeSymbols?  Seems odd.

Mike



On Oct 17, 2016, at 2:32 PM, Ben Harshbarger <notifications@github.com<mailto:notifications@github.com>> wrote:


@benharsh commented on this pull request.

________________________________

In compiler/AST/astutil.cpp<https://github.com/chapel-lang/chapel/pull/4745>:

> @@ -655,36 +679,52 @@ bool isTypeExpr(Expr* expr)
   return false;
 }

+static void pruneVisit(TypeSymbol* ts, Vec<FnSymbol*>& fns, Vec<TypeSymbol*>& types);


As I recall I made 'pruneVisitFn' a different name because pruneVisit(Symbol* sym, ...) was catching FnSymbols.

File: compiler/AST/expr.cpp [hunk 3]
```diff
@@ -4171,7 +4308,7 @@ GenRet CallExpr::codegen() {
       // Handle passing strings to externs
       //should this be else if?
       if (fn->hasFlag(FLAG_EXTERN)) {
-        if (actualType->symbol->hasFlag(FLAG_WIDE_REF) == true ||
+        if (actual->isWideRef() == true ||
```
Comment (noakesmichael) [inline compiler/AST/expr.cpp]: Trivial: Before the "equal sign" operators lined up. Now they are misaligned.

File: compiler/optimizations/bulkCopyRecords.cpp [hunk 4]
```diff
@@ -43,13 +43,38 @@ static bool isAssignment(FnSymbol* fn)
   return true;
 }
 
+std::map<Type*, bool> containsRef;
```
Comment (noakesmichael) [inline compiler/optimizations/bulkCopyRecords.cpp]: I think "containsRef" could be static.

File: compiler/optimizations/bulkCopyRecords.cpp [hunk 5]
```diff
@@ -43,13 +43,38 @@ static bool isAssignment(FnSymbol* fn)
...
+
+static bool typeContainsRef(Type* t, bool isRoot = true)
```
Comment (noakesmichael) [inline compiler/optimizations/bulkCopyRecords.cpp]: Curious.  Why a map and not a field on Type?

File: compiler/optimizations/bulkCopyRecords.cpp (hunk 4 above)
Comment (benharsh) [inline compiler/optimizations/bulkCopyRecords.cpp]: Indeed, I'll make that change.

File: compiler/passes/insertWideReferences.cpp [hunk 6]
```diff
@@ -2203,6 +2358,34 @@ static void fixAST() {
         }
         else if (Type* wide = wideRefMap.get(act->typeInfo())) {
           insertWideTemp(wide, act);
+        } else if (act->isRef()) {
```
Comment (noakesmichael) [inline compiler/passes/insertWideReferences.cpp]: I observe that the brace-style for the new else-if statements differs from existing style.
More 'note' than 'take action'.

**Pull request merged**